posts/react-compound-components/2021-5-25-React-compound-compnents.md

```diff
@@ -0,0 +1,250 @@
+
+> Context provides a way to pass data through the component tree without having to pass props down manually at every level.
+
+Now let's add context to our accordion component.
+
+```tsx
+import { createContext, useContext } from "react";
+
+// Creating the context for the Accordion.
+export const AccordionContext = createContext<{
+  activeItem: string;
+  changeSelectedItem: (item: string) => void;
+}>({ activeItem: "", changeSelectedItem: () => {} });
+
+export const useAccordionContext = () => {
+  const context = useContext(AccordionContext);
+  if (!context) {
+    throw new Error("Error in creating the context");
+  }
+  return context;
+};
+``` 
+
+After creating the context, we need to provide values to the context, it is done by **Context.Provider** element. The values refers to the data that is to be shared among all the components in the accordion. The **Provider** component accepts a value prop to be passed to consuming components that are descendants of this Provider. One Provider can be connected to many consumers.
+
+```tsx
+const Accordion: React.FC<{
+  children: ReactNode | ReactNode[];
+  className?: string;
+}> = ({ children, className }) => {
+  const [activeItem, setActiveItem] = useState("");
+
+  const changeActiveItem = useCallback(
+    (value) => {
+      if (activeItem !== value) setActiveItem(value);
+    },
+    [setActiveItem, activeItem]
+  );
+  return (
+    <AccordionContext.Provider
+      value={{ activeItem, changeSelectedItem: changeActiveItem }}
+    >
+      <div className={`accordion ${className}`}>{children}</div>
+    </AccordionContext.Provider>
+  );
+};
+
+export default Accordion;
+``` 
+
+In the accordion, we need to share the activeItem and changeSelectedItem to all the other components, so we are passing those two to the value of accordion context provider. Now We had built the root accordion component and also we have provided the values to the context. Let's build the remaining components and we are going to use the values consumed from the context and make the accordion component work as a whole.
+
+```tsx
+// AccordionItem component
+export const AccordionItem: React.FC<{
+  children: ReactNode[];
+  label: string;
+className?: string;
+}> = ({ children, label, className }) => {
+  const childrenArray = React.Children.toArray(children);
+
+  // label is used to distinguish between each accordion element.
+  // Adding the label prop to the children of accordionItem along with other props.
+  const accordionItemChildren = childrenArray.map((child) => {
+    if (React.isValidElement(child)) {
+      return React.cloneElement(child, {
+        ...child.props,
+        label
```

Comment:
Since this is more like an internal prop , you might want to name it as __cb__internal__accordion_label . So that it does not override a user prop named similarly